scene/resources/material.cpp

```diff
@@ -1619,6 +1619,11 @@ void fragment() {)";
 		code += R"(
 	// Proximity Fade: Enabled
 	float depth_tex = textureLod(depth_texture, SCREEN_UV, 0.0).r;
+)";
+		if (OS::get_singleton()->get_current_rendering_method() == "gl_compatibility") {
+			code += "	depth_tex = depth_tex * 2.0 - 1.0;";
+		}
```

Comment:
Could we make this introduced by the shader preprocessor? This way, if you convert the BaseMaterial3D to a ShaderMaterial, it'll keep working if you switch rendering methods afterwards.

(Also, I thought we already had automatic depth correction when using the Compatibiltiy rendering method.)

Reply:
> I thought we already had automatic depth correction when using the Compatibility rendering method

I thought we had this too, not sure why it doesn't work in the standard material shader. I should investigate more.

Reply:
Indeed it would be great if the depth_texture was driver agnostic, so the same shader code would work regardless of renderer. Or a preprocessor macro that converts it into 0-1 (and does nothing if Vulkan).

Reply:
I'm sorry for my previous explaination of this problem, depth is 0-1 but the ndc is different in opengl and vulkan. It's introduced in https://github.com/godotengine/godot/pull/54285, it changed to vulkan way of ndc so it caused this issue.

Visual shader also use an if here, see https://github.com/godotengine/godot/pull/74910. But we can't use this approach as Calinou mentioned.

Reply:
In 4.4+ we now have a way to do this with preprocessor defines. I made a PR: https://github.com/godotengine/godot/pull/100220.

Reply:
> In 4.4+ we now have a way to do this with preprocessor defines. I made a PR: #100220.

Seems pretty reasonable. Closing this pr now.

Reply:
> It looks like preprocessor macros don't quite work right now in the generated shaders in materials.

I suppose this can be twicked a little? Not sure it's a bug or not, but worth investigating.